File: nodejs/src/capabilities/file_version.ts
/**
 * Peek on-disk File_Version without assuming HDF5 group layout.
 * Root attribute File_Version is the adapter key.
 */
import * as h5wasm from 'h5wasm/node';
import { resolve } from 'node:path';

export const FILE_VERSION_ATTR = 'File_Version';

export class UnsupportedFileVersion extends Error {
  readonly version: string;
  constructor(version: string) {
    super(
      `No adapter registered for File_Version "${version}". Each on-disk version needs its own adapter.`,
    );
    this.name = 'UnsupportedFileVersion';
    this.version = version;
  }
}

let _readyPromise: Promise<void> | null = null;

function ensureReady(): Promise<void> {
  if (_readyPromise === null) {
    _readyPromise = h5wasm.ready.then(() => undefined);
  }
  return _readyPromise;
}

function normPath(p: string): string {
  return resolve(p).replace(/\\/g, '/');
}

function attrVersion(attrs: Record<string, h5wasm.Attribute>): string {
  const attr = attrs[FILE_VERSION_ATTR];
  if (attr == null || attr.value == null) return '1.0';
  const v = attr.value;
  const s = String(ArrayBuffer.isView(v) && 'length' in v ? (v as ArrayLike<unknown>)[0] : v).trim();
  return s || '1.0';
}

/** Read only the root File_Version attribute. Does not walk groups. */
export async function peekFileVersion(path: string): Promise<string> {
  await ensureReady();
  const f = new h5wasm.File(normPath(path), 'r');
  try {
    return attrVersion(f.attrs as Record<string, h5wasm.Attribute>);
  } finally {
    f.close();
  }
}

File: python/src/machine_config/capabilities/file_version.py
"""Peek on-disk File_Version without assuming HDF5 group layout.

Root attribute ``File_Version`` is the adapter key. Everything else (groups,
attribute names, casting) lives in the matching File_Version adapter.
"""
from __future__ import annotations

from pathlib import Path

import h5py

FILE_VERSION_ATTR = "File_Version"


class UnsupportedFileVersion(ValueError):
    def __init__(self, version: str) -> None:
        self.version = version
        super().__init__(
            f'No adapter registered for File_Version "{version}". '
            "Each on-disk version needs its own adapter."
        )


class MissingRequiredGroup(KeyError):
    """A required HDF5 group is absent in an otherwise valid file."""
    def __init__(self, path: str) -> None:
        self.path = path
        super().__init__(f"Required HDF5 group missing: {path}")


def peek_file_version(path: str | Path) -> str:
    """Read only the root File_Version attribute. Does not walk groups."""
    with h5py.File(path, "r") as f:
        raw = f.attrs.get(FILE_VERSION_ATTR, "1.0")
        v = str(raw).strip()
        return v or "1.0"
